### DISCUSSION TITLE
Twin.Macro with Storybook and SVGR Loader

### DISCUSSION THREAD

**Discussion by LouisJackson** [ben-rogerson/twin.macro#545] [Q&A] [answered]
Hello I'm using twin.macro with Styled-Components within Storybook and, I'm using SVGR Loader for svgs.

When I'm adding tw and css props to my SVG component it does work normally, unless I'm using the component with a dynamic name.

```.tsx
import React from 'react';
import tw, { TwStyle } from 'twin.macro';
import ArrowRightIcon from '../../../assets/arrow-right.svg';

const IconVariants = {
  'arrow-right': ArrowRightIcon,
}

interface Props extends React.SVGAttributes<SVGElement> {
  variant?: keyof typeof IconVariants;
  extraCss: string | TwStyle | TwStyle[];
}

const Icon: React.FC<Props> = ({ variant = 'arrow-right', extraCss }) => {
  const IconComponent = IconVariants[variant];

  return (
    <IconComponent tw="inline fill-current stroke-current" css={extraCss} />
  ); // <-- doesn't work
  return (
    <ArrowRightIcon tw="inline fill-current stroke-current" css={extraCss} />
  ); // <-- work
};

export default Icon;

```

Here are the errors I get in Storybook, when I add just the `tw` property and both `tw` and `css` property.

<img width="894" alt="Screenshot 2021-10-06 at 11 24 48" src="https://user-images.githubusercontent.com/8021023/136176559-e8cf287b-db99-4181-9984-20324481debb.png">

<img width="899" alt="Screenshot 2021-10-06 at 11 24 38" src="https://user-images.githubusercontent.com/8021023/136176583-43f61fc9-1dc6-4712-9785-188391ed466b.png">

I wonder if it is a twin.macro bug or if I'm doing things wrong.

Thanks for the help!

**Comment by ben-rogerson** [comment, marked as answer]
Hey there!
Could you also share an example of your `IconVariants` component so I can see how you're selecting the icon (and passing the props to it)?

**Reply by LouisJackson** [reply]
Updated my post to include IconVariants.

**Reply by LouisJackson** [reply]
Hey @ben-rogerson ,
Did you have the chance to wrap your head around iut? Should I submit a bug issue?
Thanks in advance! 🙏

**Reply by ben-rogerson** [reply]
Hey Louis, could you try passing the className prop to the svg, like this:

```js
const ArrowRightIcon = ({ className }) => <svg className={className}>...</svg> 
```

**Reply by LouisJackson** [reply]
Hey Ben, the issue remains.

**Reply by ben-rogerson** [reply]
Whats your webpack config for svg loader? 
I'd have to recreate an example to get to the bottom of this. 
Providing me an example repo would be ideal though :)

**Reply by LouisJackson** [reply]
[Here's a repo](https://github.com/LouisJackson/test-twin-svgr-storybook) to reproduce the problem based on the Twin TS StyledComponents Storybook template you made.
Changes have been made to `.storybook/main.js` file to include the SVGR Loader which works fine until you add a `tw` prop to the SVG. Hope this helps you get to the bottom of this.

**Reply by LouisJackson** [reply]
@ben-rogerson did you have the chance to give it a look?

**Reply by ben-rogerson** [reply]
Twin hoists the styled component outside of the component when the tw prop is used and so loses it's scope

After babel runs, it looks like this:

```js
import React from "react";
import ArrowRightIcon from "../assets/arrow-right.svg";
const IconVariants = { "arrow-right": ArrowRightIcon };

var _StyledIconComponent = _styled(IconComponent).withConfig({ // < IconComponent is no longer in scope
  displayName: "pages___StyledIconComponent",
  componentId: "dl4jt5-0"
})({ "borderWidth": "1px" });

const Icon = ({ variant = "arrow-right" }) => {
  const IconComponent = IconVariants[variant];
  return <_StyledIconComponent data-tw="border" />;
};

export default Icon;
```

The workaround is to convert `IconComponent` to an outside function, something like this:

```js
// Icon.tsx
import React from "react";
import tw, { TwStyle } from "twin.macro";
import ArrowRightIcon from "../assets/arrow-right.svg";

const IconVariants: { [key: string]: React.FC } = {
  "arrow-right": ArrowRightIcon,
};

interface Props extends React.SVGAttributes<SVGElement> {
  variant: keyof typeof IconVariants;
  // extraCss: string | TwStyle | TwStyle[];
}

const IconComponent = ({ variant, ...rest }: {
  variant: keyof typeof IconVariants;
}) => IconVariants[variant](rest);

const Icon: React.FC<Props> = ({ variant = "arrow-right" }) => (
  <IconComponent tw="border" variant={variant} />
);

export default Icon;
```

Alternatively, you could return `IconVariants` directly:

```js
// Icon.tsx
import React from "react";
import ArrowRightIcon from "../assets/arrow-right.svg";

const IconVariants {
  "arrow-right": ArrowRightIcon,
};

interface Props extends React.SVGAttributes<SVGElement> {
  variant: keyof typeof IconVariants;
}

const Icon: React.FC<Props> = (props) =>
  IconVariants[props.variant || "arrow-right"](props);

export default Icon;
``` 

and:

```js
// Icon.stories.tsx
import React from "react";
import { Story } from "@storybook/react/types-6-0";
import Icon from "./Icon";
import "twin.macro";

export default { title: "Icon", component: Icon };

export const Default: Story = (props) => (
  <Icon variant="arrow-right" tw="border" {...props} />
);
```

**Reply by LouisJackson** [reply]
thanks a lot! makes sense and will use the first workaround for now.
